delegate-app/Main.hs

```diff
@@ -1,49 +1,133 @@
+
+-- Hydra imports
+import Hydra.Network
 
 -- HydraAuction imports
 import HydraAuction.Delegate (
   DelegateEvent (Start),
-  DelegateInput (DelegateEvent),
+  DelegateInput (DelegateEvent, FrontendRequest),
   DelegateRunnerT,
   delegateStep,
   execDelegateRunnerT,
  )
+import HydraAuction.Delegate.Server (
+  DelegateError (FrontendNoParse),
+  DelegateServerConfig (DelegateServerConfig, dlgt'host, dlgt'port, dlgt'tick),
+  DelegateServerLog (DelegateError, DelegateOutput, FrontendConnected, FrontendInput, Started),
+  MonadTracer (trace),
+  ServerAppT,
+  TracerT,
+  askTracer,
+  dlgt'tick,
+  runWithTracer',
+ )
 
-eventsProducer :: Bool -> DelegateRunnerT IO (Maybe DelegateEvent)
+eventsProducer :: Monad m => Bool -> DelegateRunnerT m (Maybe DelegateEvent)
 eventsProducer thisIsFirstRun =
   if thisIsFirstRun
     then return $ Just Start
     else return Nothing
 
-consumer :: Maybe DelegateInput -> DelegateRunnerT IO ()
-consumer mInput = case mInput of
-  Just input -> do
-    delegateResponse <- delegateStep input
-    -- FIXME: send response to client
-    liftIO $
-      putStrLn $
-        "Delegate responses for input: " <> show delegateResponse
-  Nothing -> return ()
+delegateServerApp ::
+  (MonadIO m, MonadTracer DelegateServerLog m) =>
+  Int ->
+  ServerAppT m
+delegateServerApp tick pending = do
+  connection <- liftIO $ acceptRequest pending
+  trace FrontendConnected
+  -- FIXME: we will probably have to do this in parallel, but how can we enqueue in the same
+  -- state? probably have to remodel this
+  execDelegateRunnerT $ flip evalStateT True $ mkRunner tick connection
+
+-- FIXME: remove concrete monads and make this tagless final (needs MonadDelegate)
+mkRunner ::
+  (MonadIO m, MonadTracer DelegateServerLog m) =>
+  Int ->
+  Connection ->
+  StateT Bool (DelegateRunnerT m) ()
+-- FIXME: we need to abort at some point but this doesn't seem
+-- to be implemented yet so we just go on
+mkRunner tick con = forever $ do
+  isFirstRun <- get
+  _ <- lift $ do
+    let encodeSend = liftIO . sendTextData con . encode
+    liftIO (receiveData con)
+      >>= ( \case
+              Nothing -> lift $ trace (DelegateError FrontendNoParse)
+              Just req -> do
+                lift . trace $ FrontendInput req
+                step <- delegateStep (FrontendRequest req)
+                lift $ traverse_ (trace . DelegateOutput) step
+                encodeSend step
+          )
+        -- yes, this is not nice, but hlint wants it so badly
+        . decode
 
-delegateTick :: Int
-delegateTick = 1_000
+    -- FIXME: more elaborate logic for eventsProducer
+    mdelegateEvent <- eventsProducer isFirstRun
+    whenJust (DelegateEvent <$> mdelegateEvent) $
+      delegateStep >=> encodeSend
+
+  liftIO $ threadDelay tick
+  put False
+
+runDelegateServer ::
+  DelegateServerConfig ->
+  TracerT DelegateServerLog IO ()
+runDelegateServer conf = do
+  trace (Started $ dlgt'port conf)
+  let fallback :: Application
```

Comment:
What does this code do?

Reply:
Like it is some HTTP-specific stuff. Hydra just uses `runServer` and that seems enough.

Hydra code is probably over complicated over what we need.

Reply:
in the docs it explicitly said you should not use `runServer`, that's why I didn't, this basically offers a fallback if the frontend tries to communicate via HTTP instead of websockets

Reply:
https://hackage.haskell.org/package/websockets-0.12.7.3/docs/Network-WebSockets.html#v:runServer

Reply:
I will change it to `runServer`